### PR TITLE
Feature vs4 refactor distance class

### PR DESCRIPTION
**Pull request by JulianvDoorn**
## Will do:
In this slice the interface definitions will be explicitly noted. So that external modules will not be required to invoke internal functions in order to filter the requested data.

~~`getSensorData()`~~
~~`medianFilter(int a[], int size)`~~
~~`Distance (hwlib::pin_in HCSR04_pin_in, hwlib::pin_out HCSR04_pin_out, hwlib::pin_in laser_rx_pin)`~~
~~`Distance::getDistance(Enum Sensor, Enum Scale)`~~
~~`HCSR04::getDistance(hwlib::pin_in HCSR04_pin_in, hwlib::pin_out HCSR04_pin_out)`~~
~~`LIDARmini::getDistance()`~~

`getDistance(): Integer`
Applies the median filter to the previous 10 values and the newly read raw value from the sensor specified by setSensor and returns the calculated value in cms. 

`setSensor(Sensor: Auto | Lidar | Ultrasonic)`
Sets the sensor specifier to the sensor specified in the parameter.

`setUnit(Unit: cm | inch)`
Does nothing

`setFilter(Filter: Median | Kalman, Size: Integer)`
Does nothing

#### Note
Some interface functions have been removed from the vertical slice definition. They are not removed from the code, but they are internal components and are not necessary to formulate vertical slices.

Vertical slice: https://trello.com/c/k4wpph8q/49-vs4-discrete-module-interface
Backlog items in this branch:
- https://trello.com/c/gNA68Gfy/59-vs4-refactor-distance-class-2
- https://trello.com/c/DwfI5WjZ/51-vs4-setsensor-1

### CODE REVIEW

File: src/HCSR04.cpp [hunk 1]
```diff
@@ -19,22 +18,25 @@ int HCSR04::getDistance() {
     hwlib::wait_ms(10);
     trigger_pin.set(0);
 
+    // Get current time
```
Comment (vasilishomegas) [inline src/HCSR04.cpp]: Inline comments like this:
///< Comment
This goes for elsewhere too.

File: src/HCSR04.cpp [hunk 2]
```diff
@@ -19,22 +18,25 @@ int HCSR04::getDistance() {
     hwlib::wait_ms(10);
```
Comment (vasilishomegas) [inline src/HCSR04.cpp]: I believe this delay can be 1000 times shorter.
Also, the ```hwlib::wait_ms(2);``` is unnecessary.

File: src/HCSR04.cpp [hunk 3]
```diff
@@ -19,22 +18,25 @@ int HCSR04::getDistance() {
...
-
-    while (!received) {
-        received = echo_pin.get();
-    }
-    while (received) {
-        received = echo_pin.get();
+    // If the first pulse hasn't been received yet, wait until first pulse is found.
+    if (!received) {
+        while (!(received = echo_pin.get()))
+            ;
```
Comment (vasilishomegas) [inline src/HCSR04.cpp]: Use this:
```
while(condition) {
}
```
instead of the semicolon.

File: src/HCSR04.cpp [hunk 4]
```diff
@@ -19,22 +18,25 @@ int HCSR04::getDistance() {
...
-    // Get current time
-    int duration = (int)hwlib::now_us();
-
-    while (!received) {
-        received = echo_pin.get();
-    }
-    while (received) {
-        received = echo_pin.get();
+    // If the first pulse hasn't been received yet, wait until first pulse is found.
+    if (!received) {
```
Comment (vasilishomegas) [inline src/HCSR04.cpp]: This is an unnecessary extra check. Just check if the signal goes high directly.

File: src/HCSR04.cpp [hunk 5]
```diff
@@ -19,22 +18,25 @@ int HCSR04::getDistance() {
...
+        while (!(received = echo_pin.get()))
+            ;
     }
-
-    // Calculate distance
-    duration = (int)((hwlib::now_us() - duration) - 400);
-    int distance = (int)((duration / 2) / 29.1);
+    // While the first pulse has just occurred, keep on looping until echo_pin returns false.
+    while ((received = echo_pin.get()))
+        ;
```
Comment (vasilishomegas) [inline src/HCSR04.cpp]:
```
while(condition) {
}
```

File: src/HCSR04.cpp [hunk 6]
```diff
@@ -19,22 +18,25 @@ int HCSR04::getDistance() {
...
+
+    // Calculate distance by converting duration to centiseconds,
+    // multiplying with the speed of sound (343m/s) and dividing this roundtrip distance by two.
+    duration = hwlib::now_us() - duration;
+    int distance = (int)((duration / 29.1) / 2);
 
     // Return distance, when it exceeds it's maximum range; it returns 400.
```
Comment (vasilishomegas) [inline src/HCSR04.cpp]: I don't know if just saying it's 400 if you don't know is the right move.

File: src/HCSR04.cpp [hunk 7]
```diff
@@ -19,22 +18,25 @@ int HCSR04::getDistance() {
...
     }
-
-    // Calculate distance
-    duration = (int)((hwlib::now_us() - duration) - 400);
-    int distance = (int)((duration / 2) / 29.1);
+    // While the first pulse has just occurred, keep on looping until echo_pin returns false.
+    while ((received = echo_pin.get()))
+        ;
+
+    // Calculate distance by converting duration to centiseconds,
```
Comment (vasilishomegas) [inline src/HCSR04.cpp]: centiseconds?

File: src/distance.hpp [hunk 8]
```diff
@@ -66,7 +84,7 @@ class Distance {
      * @param[in]     Distance::Scale       The scale in which the distance has to be returned
      * @return        inches
      */
-    int convertToInches(int centimeters);
+    static int convertToInches(int centimeters);
```
Comment (vasilishomegas) [inline src/distance.hpp]: Doxygen mentions two parameters, but function only takes one. It seems like the scale parameter is unnecessary.

File: src/hardware_uart.hpp [hunk 9]
```diff
@@ -195,6 +174,18 @@ class UARTConnection {
      * @return char
      */
     inline char receiveByte();
+
+    void putc(char c) override {
```
Comment (vasilishomegas) [inline src/hardware_uart.hpp]: No doxygen, but maybe this needs to be addressed in the UART library.

File: src/io_stream.hpp [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+#include "wrap-hwlib.hpp"
+
+/**
+ * @brief Interface inheriting hwlib::istream and hwlib::ostream
+ *
+ * @details
+ * The purpose of this interface is to merge the istream and ostream.
+ */
+class IOStream : public hwlib::istream, public hwlib::ostream {
+    virtual void putc(char c) = 0;
```
Comment (vasilishomegas) [inline src/io_stream.hpp]: No Doxygen

File: src/io_stream.hpp [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+    virtual bool char_available() = 0;
```
Comment (vasilishomegas) [inline src/io_stream.hpp]: No Doxygen

File: src/io_stream.hpp [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+
+    virtual char getc() = 0;
```
Comment (vasilishomegas) [inline src/io_stream.hpp]: Doxygen

**Review by vasilishomegas — CHANGES_REQUESTED** [review]

File: src/io_stream.hpp (hunk 10 above)
Comment (Dylrak) [reply src/io_stream.hpp]: Added doxygen for entire IOstream class

File: src/lidar_mini.cpp [hunk 13]
```diff
@@ -38,15 +38,13 @@ std::array<char, 9> LIDARmini::getWantedRegisters(char registerSetByte) {
                     wantedRegisters[1] = 0x59;
                     registerCounter += 2;
                 }
-
-                for (int i = 1; i < 8; i++) {
-                    if (((registerSetByte >> i) & 1) != 0) {
-                        wantedRegisters[registerCounter] = dat[i];
+                for (int i = 6; i >= 0; i--) { // Loop for 7 more times to fill in the remaining data package in the array.
```
Comment (vasilishomegas) [inline src/lidar_mini.cpp]: Inline comments:
///< Comment

File: src/HCSR04.cpp [hunk 14]
```diff
@@ -13,31 +13,21 @@ HCSR04::HCSR04(hwlib::pin_out &trigger_pin, hwlib::pin_in &echo_pin) : trigger_p
...
 int HCSR04::getDistance() {
     trigger_pin.set(0);
-    hwlib::wait_ms(2);
     trigger_pin.set(1);
-    hwlib::wait_ms(10);
+    hwlib::wait_us(10);
     trigger_pin.set(0);
-
-    // Get current time
+    /// Wait for first signal from echo pin
```
Comment (vasilishomegas) [inline src/HCSR04.cpp]: ///< Comment

File: src/HCSR04.cpp [hunk 15]
```diff
@@ -13,31 +13,21 @@ HCSR04::HCSR04(hwlib::pin_out &trigger_pin, hwlib::pin_in &echo_pin) : trigger_p
...
-    // While the first pulse has just occurred, keep on looping until echo_pin returns false.
-    while ((received = echo_pin.get()))
-        ;
-
-    // Calculate distance by converting duration to centiseconds,
-    // multiplying with the speed of sound (343m/s) and dividing this roundtrip distance by two.
+    /// Calculate distance by calculating the delta duration
     duration = hwlib::now_us() - duration;
-    int distance = (int)((duration / 29.1) / 2);
+    int distance = (int)(duration / 58);
```
Comment (vasilishomegas) [inline src/HCSR04.cpp]: Is (int) still necessary? Everything is an int already, although duration is a long long int. If it doesn't work right without it, it's fine.

File: test/hcsr04_test_bench.hpp [hunk 16]
```diff
@@ -204,24 +198,18 @@ TEST_CASE("Test HCSR04TestBench Blast Time") {
...
-        /// If the elapsed time is more than one second, fail this test.
-        // REQUIRE(elapsedTime <= 1000000);
     }
+    long long int startTime = hwlib::now_us();
+    long long int elapsedTime = 0;
     /// After this while loop, echoPin.get() returns true.
     /// While-loop until echoPin.get() returns false
-    while (echoPin.get() == true)
-        ;
+    while (echoPin.get() == true) {
```
Comment (vasilishomegas) [inline test/hcsr04_test_bench.hpp]: ```== true``` is unneccesary

File: src/lidar_mini.cpp [hunk 17]
```diff
@@ -21,30 +21,35 @@ std::array<char, 9> LIDARmini::getWantedRegisters(char registerSetByte) {
...
             char dat[8];
-
             for (int i = 0; i < 8; i++) {
+                while (!uart.available()) {
+                }
                 uart >> dat[i];
             }
-
+            for (int i = 0; i < 8; i++) {
+                hwlib::cout << hwlib::hex << (int)dat[i] << "\r\n";
```
Comment (vasilishomegas) [inline src/lidar_mini.cpp]: You don't need the \r, just fix your line endings in PuTTY (I assume that is what you use).

**Review by vasilishomegas — CHANGES_REQUESTED** [review]

**Review by vasilishomegas — APPROVED** [review]

**Pull request merged**